Buffer overflows

When compiling the latest master on macOS High Sierra, clang points out a couple buffer overflows in your code:

```
tcp_subr.c:341:4: warning: '__builtin___strncpy_chk' will always overflow destination buffer [-Wbuiltin-memcpy-chk-size]
                        strncpy(addr.sun_path,path,UNIX_PATH_MAX);
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/secure/_string.h:124:3: note: expanded from macro 'strncpy'
                __builtin___strncpy_chk (dest, __VA_ARGS__, __darwin_obsz (dest))
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Here, it looks like you're copying `UNIX_PATH_MAX` bytes to the `sun_path` field of a `struct sockaddr_un`, and in tcp2unix.h you've defined `UNIX_PATH_MAX` as 108, which as far as I can tell is correct for Linux systems, but based on cursory research, it seems that different operating systems use different sizes for this field, down to as small as 92 bytes. On my macOS system, the `sun_path` field's length is 104 bytes.

```
vder_packet.c:207:5: warning: '__builtin___memcpy_chk' will always overflow destination buffer [-Wbuiltin-memcpy-chk-size]
                                memcpy(foot, footprint(packet), sizeof(struct iphdr) + 8);
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/secure/_string.h:62:3: note: expanded from macro 'memcpy'
                __builtin___memcpy_chk (dest, __VA_ARGS__, __darwin_obsz0 (dest))
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

In this case, I think the problem is that you're allocating the variable `foot` based on the size of a pointer rather than the size of the struct that the pointer points to.

--- a/src/lib/libvdesnmp.c
+++ b/src/lib/libvdesnmp.c
@@ -261,7 +261,9 @@ int counters_parse(void){
 						PORTDOWN(curport-1);
 					} else if( strncmp(portstatus, "ACTIVE", 6) == 0 ){
 						PORTUP(curport-1);
-						strncpy(pl->desc, portdesc, strlen(portdesc)-1);
+						/* Ensure string is null-terminated before strcopy */
+						portdesc[DESC_MAXLEN - 1] = (char)0;
+						strcpy(pl->desc, portdesc);
 					}
 
 					inpkts = inbytes = outpkts = outbytes = 0;

--- a/src/vde_cryptcab/cryptcab.c
+++ b/src/vde_cryptcab/cryptcab.c
@@ -91,8 +91,6 @@ isvalid_crc32(unsigned char *block, int len)
 
 int data_encrypt_decrypt(unsigned char *src, unsigned char *dst, int len, unsigned char *key, unsigned char *iv)
 {
-	int tlen, olen;
-
 	if (encryption_disabled){
 		memcpy(dst,src,len);
 		return len;

--- a/src/vde_cryptcab/cryptcab.c
+++ b/src/vde_cryptcab/cryptcab.c
@@ -99,6 +99,7 @@ int data_encrypt_decrypt(unsigned char *src, unsigned char *dst, int len, unsign
 	wc_Chacha_SetIV(&ctx, iv, CHACHA_IV_BYTES);
 	if (wc_Chacha_Process(&ctx, dst, src, len) == 0)
 		return len;
+    return -1;
 }
 
 

--- a/src/vde_cryptcab/cryptcab.h
+++ b/src/vde_cryptcab/cryptcab.h
@@ -36,6 +36,7 @@
 #include <config.h>
 #include <vde.h>
 #include <vdecommon.h>
+#include <limits.h>
 
 
 #define PORTNO 7667

--- a/src/vde_l3/vde_l3.c
+++ b/src/vde_l3/vde_l3.c
@@ -428,12 +428,12 @@ static void swap_macaddr(uint8_t addr1[], uint8_t addr2[])
 /* 
  * Swap src/dst ip addresses at given mem addresses
  */
-static void swap_ipaddr(uint32_t *addr1, uint32_t *addr2)
+static void swap_ipaddr(void *addr1, void *addr2)
 {
-	uint32_t tmp;
-	memcpy(&tmp,addr1,4);
+	uint8_t tmp[sizeof(uint32_t)];
+	memcpy(tmp,addr1,4);
 	memcpy(addr1,addr2,4);
-	memcpy(addr2,&tmp,4);
+	memcpy(addr2,tmp,4);
 }
 
 
@@ -535,7 +535,7 @@ size_t arp_reply(struct vde_iface *oif, struct vde_buff *vdb)
 	ah =(struct arp_header *)iphead(vdb);
 	ah->opcode = htons(ARP_REPLY);
 	swap_macaddr(ah->s_mac, ah->d_mac);
-        memcpy(ah->s_mac, oif->mac,6);
+	memcpy(ah->s_mac, oif->mac,6);
 	swap_ipaddr(&(ah->s_addr), &(ah->d_addr));
 
 	return(raw_send(oif,vdb));
@@ -1460,7 +1460,7 @@ int main(int argc, char *argv[])
 	i=0;
 	while (vif) {
 		pfd[i].fd = vde_datafd(vif->vdec);
-     		pfd[i++].events=POLLIN | POLLHUP;
+		pfd[i++].events=POLLIN | POLLHUP;
 		vif = vif->next;
 	}
 	npfd = numif;

--- a/src/vde_l3/vde_l3.h
+++ b/src/vde_l3/vde_l3.h
@@ -1,5 +1,5 @@
-#ifndef _VDE_L3_H_
-#define _VDE_L3_H__
+#ifndef VDE_L3_H
+#define VDE_L3_H
 /* pfifo.c */
 int pfifo_enqueue(struct vde_buff *vdb, struct vde_iface *vif);
 int pfifo_dequeue(struct vde_iface *vif);
@@ -49,4 +49,4 @@ uint32_t ascii2ip(char *c);
 uint32_t valid_nm(uint32_t nm);
 uint32_t ascii2nm(char *c);
 int ip_input(struct vde_buff *vdb);
-#endif /* _VDE_L3_H__ */
+#endif /* VDE_L3_H */

--- a/src/vde_router/vde_router.c
+++ b/src/vde_router/vde_router.c
@@ -579,7 +579,7 @@ static void show_filter(int fd, struct vder_filter *filter)
 		snprintf(source, 10, "eth%d", filter->src_iface->interface_id);
 	}
 	if (filter->tos >= 0) {
-		snprintf(tos, 10, "tos %d", filter->tos);
+		snprintf(tos, 10, "tos %3d", (uint8_t)filter->tos);
 	}
 	if (filter->proto > 0) {
 		proto_name(filter->proto, proto);

--- a/src/vde_router/vder_packet.c
+++ b/src/vde_router/vder_packet.c
@@ -202,7 +202,7 @@ void vder_packet_recv(struct vder_iface *vif, int timeout)
 			} else {
 				struct iphdr *hdr = iphead(packet);
 				uint32_t sender = hdr->saddr;
-				uint8_t foot[sizeof(hdr) + 8];
+				uint8_t foot[sizeof(struct iphdr) + 8];
 
 				memcpy(foot, footprint(packet), sizeof(struct iphdr) + 8);
 				if (vder_ip_decrease_ttl(packet)) {

